faction.cpp

```diff
@@ -301,144 +310,366 @@ AString Faction::FactionTypeStr()
 	return temp;
 }
 
-void PadStrEnd(AString& str, int len) {
-	int size = str.Len();
-	for (int i = len; i > size; i--) str += AString(" ");
-}
+vector<FactionStatistic> Faction::compute_faction_statistics(Game *game, size_t **citems) {
+	vector<FactionStatistic> stats;
+	// To allow testing, just return an empty vector if we don't have any item arrays inbound
+	if (!citems) return stats;
 
-void Faction::WriteFactionStats(ostream& f, Game *pGame, int ** citems) {
-	f << ";Item                                      Rank  Max        Total\n";
-	f << ";=====================================================================\n";
-	for (int i = 0; i < NITEMS; i++)
-	{
+	auto myfaction = this->num - 1; // offset by 1 since the citems array are 0-based and factions are 1-based.
+
+	for (auto i = 0; i < NITEMS; i++) {
 		if (ItemDefs[i].type & IT_SHIP) continue;
+		if(citems[myfaction][i] == 0) continue;
+		size_t place = 1;
+		size_t max = 0;
+		size_t total = 0;
+
+		for (int pl = 0; pl < game->factionseq; pl++) {
+			if (citems[pl][i] > citems[myfaction][i]) place++;
+			if (max < citems[pl][i]) max = citems[pl][i];
+			total += citems[pl][i];
+		}
```

Comment:
Since you had already iterated all items for all units for all factions while building the citems array, there was no need to do so again here, so I just rewrote this to use the data you'd already collected :p

Reply:
Also, it now returns it as a vector so I can output it in textual and json formats as needed.